cloudinit/url_helper.py

```diff
@@ -394,31 +517,94 @@ def wait_for_url(
 
     A value of None for max_wait will retry indefinitely.
     """
-    start_time = time.time()
-
-    def log_status_cb(msg, exc=None):
-        LOG.debug(msg)
 
-    if status_cb is None:
-        status_cb = log_status_cb
+    def default_sleep_time(_, loop_number: int):
+        return int(loop_number / 5) + 1
 
     def timeup(max_wait, start_time):
+        """Check if time is up based on start time and max wait"""
         if max_wait is None:
             return False
         return (max_wait <= 0) or (time.time() - start_time > max_wait)
 
-    loop_n = 0
-    response = None
-    while True:
-        if sleep_time_cb is not None:
-            sleep_time = sleep_time_cb(response, loop_n)
+    def read_url_handle_response(response, url):
+        """Map requests response code/contents to internal "UrlError" type"""
+        if not response.contents:
+            reason = "empty response [%s]" % (response.code)
+            url_exc = UrlError(
+                ValueError(reason),
+                code=response.code,
+                headers=response.headers,
+                url=url,
+            )
+        elif not response.ok():
+            reason = "bad status code [%s]" % (response.code)
+            url_exc = UrlError(
+                ValueError(reason),
+                code=response.code,
+                headers=response.headers,
+                url=url,
+            )
         else:
-            sleep_time = int(loop_n / 5) + 1
+            reason = ""
+            url_exc = None
+        return (url_exc, reason)
+
+    def read_url_handle_exceptions(
+        url_reader_cb, urls, start_time, exc_cb, log_cb
+    ):
+        """Execute request, handle response, optionally log exception"""
+        reason = ""
+        url = None
+        try:
+            url, response = url_reader_cb(urls)
+            url_exc, reason = read_url_handle_response(response, url)
+            if not url_exc:
+                return (url, response)
+        except UrlError as e:
+            reason = "request error [%s]" % e
+            url_exc = e
+        except Exception as e:  # yikes, do we really think this is necessary?
```

Comment:
Why the yikes? If everything doesn't get caught by UrlError, we'd want a broader exception to catch anything else.

Reply:
Oops. That's a note-to-self that slipped past my final review, agreed.